### DISCUSSION TITLE
Non-functioning spi/i2c / general confusion about ports / ESP32-D0WDQ6-Vx

### DISCUSSION THREAD

**Discussion by PsuFan** [micropython/micropython#9503] [Hardware & Peripherals] [answered]
Hey guys, can you me a rundown on the basic differences between esp32 ports or which port I should be using for the ESP32-D0WDQ6-Vx chip? D2WD? I have some v3 TTGO chips that the i2c screen does not respond when running any recent firmware. If I run the generic idf-v3 1.14 the screen runs fine. Currently having an issue with a ESP32-WROOM-32D (D0WDQ6-V1 I believe) which the spi isn't connecting (hopefully, haven't confirmed its not a circuit or chip issue yet).

**Comment by davefes** [comment]
I recall some major changes to I2C ... are you doing something like:
```py
i2c = SoftI2C(scl=Pin(22), sda=Pin(21), freq=400000)
```

ie Soft versus whatever was used before.

Likewise, are you using a Soft or Hard SPI bus?

**Reply by PsuFan** [reply]
I did get a depreciation warning with i2c on v1.14 and tried SoftI2C but there was no change. I don’t know the difference between soft and hard. It’s a direct connection on 22/21 to a gpio expander mcp23017

**Reply by davefes** [reply]
Firstly, I wouldn't mess-around with a old image.
Pin 22 and 21 are just an example that I use to talk to a ADS1115 ADC chip.  You could pick some other GPIO.

Soft = software, hard = hardware.  The ESP32 supports both for SPI and I2C
[https://docs.micropython.org/en/latest/esp32/quickref.html](url) 
Maybe, search the old forum as I thought one was recommended over the other for I2C

**Reply by PsuFan** [reply]
I don’t have a choice for the TTGO, the spi display is 100% confirmed not working on multiple chips on several of the latest firmwares. Flashing 1.14 with the exact same program and the screens work fine. Something definitely changed in the new versions or has to do with the idf4? Hopefully @jimmo or @robert-hh can explain.

**Comment by jimmo** [comment]
@PsuFan the ESP32-D0WDQ6 should use the "generic" firmware.

The only exception for the D0WD family is the ESP32-D0WDR2-V3 which can optionally use the "generic spiram" firmware.

`machine.I2C` (i.e. "hardware i2c") uses the i2c peripheral in the esp32 chip. `machine.SoftI2C` implements i2c in software (i.e. the micropython firmware toggles the pin in a loop). The main advantage of using the hardware i2c is that it can run at a higher bit rate.

On other ports, the software i2c has an additional benefit that it can work on any GPIO pin. But on ESP32, the hardware peripherals are very flexible and can generally run on any pin. (This is not true on other microcontrollers).

(Exactly the same story for SPI).

**Reply by PsuFan** [reply]
hey, speak of the devil. You posted while I was posting. Thanks for the info. Any idea why spi stopped working in newer images?

**Reply by PsuFan** [reply]
@jimmo the generic spiram does not run on "Chip is ESP32-D0WDQ6-V3 (revision 3)". What is in the LILYGO_TTGO version? This is the chip I am using: https://www.amazon.com/gp/product/B07X1W16QS/

Finally figured out the spi issue, the example I was given didn't specify a miso pin... which SoftSPI complained about. Previous images must have assumed 17, recent versions must not.

**Reply by davefes** [reply]
Hmmm the generic spiram V1.19.1 runs on both of my D0WDQ6-V3.  What doesn't work or what errors do you get?

**Reply by PsuFan** [reply]
```
ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:4540
ho 0 tail 12 room 4
load:0x40078000,len:12788
load:0x40080400,len:4176
entry 0x40080680
E (651) psram: PSRAM ID read error: 0xffffffff
E (651) spiram: SPI RAM enabled but initialization failed. Bailing out.
E (652) spiram: SPI RAM not initialized
Re-enable cpu cache.

abort() was called at PC 0x400d3ea6 on core 0

Backtrace:0x400964a9:0x3ffe3b60 0x40096a99:0x3ffe3b80 0x4009a355:0x3ffe3ba0 0x400d3ea6:0x3ffe3c10 0x400d3ef7:0x3ffe3c30 0x400825ee:0x3ffe3c50 0x40081e1b:0x3ffe3c70 0x40078f5c:0x3ffe3c90 |<-CORRUPTED


ELF file SHA256: 46bca36b7d6020a6

Rebooting...

**Comment by PsuFan** [comment]
I have confirmed in Arduino that the i2c mcp23017 is fully functional. However .scan in micropython returns []

```
>>> i2c = I2C(sda=Pin(21), scl=Pin(22))
>>> i2c.scan()
[]

**Reply by davefes** [reply]
Try adding adding freq=100000 as per my posting, ie a slower datarate

Also, looking at your error log the first line looks odd.  I'll re-boot one of mine later today but I would not expect a date 3 years in the past.

**Reply by PsuFan** [reply]
I tried 400,000 because that's what I am using in Arduino which was no change but I got it working with the "I2C(0, ...". See below, posted at same time...

**Comment by PsuFan** [comment, marked as answer]
Ugh this was as simple as needing a zero parameter. Not quite sure what these are for but they are board specific.
```
>>> i2c = I2C(0, sda=Pin(21), scl=Pin(22))
>>> i2c.scan()
[32]

**Reply by davefes** [reply]
According the docs you can define the pins for a hardware I2C, I am surprised.

**Reply by alexxus** [reply]
I had the same problem on the Wemos Lolin32 Lite. After adding the zero (0) as the first parameter, I2C started to work.

Can someone explain, what this zero (0) stands for and why it has to be set as the first parameter? Is it documented somewhere?

**Reply by robert-hh** [reply]
The first argument is the device ID, and that is documented in the documentation: https://docs.micropython.org/en/latest/library/machine.I2C.html?highlight=machine%20i2c#machine.I2C

**Reply by alexxus** [reply]
Thank you very much, @robert-hh, for pointing me to the right place in the documentation.

It slowly starts to make sense to me now.

I was following the official tutorial on the micropython website, which was about using the ssd1306 OLED display.
Sadly, there was no information about this first parameter. Too bad for all the newbies, who didn't read the I2C documentation in the first place 😐

**Reply by robert-hh** [reply]
That sections uses the ESP8266, which does not need the ID argument, since there is only one I2C bus. Maybe that should be changed for consistency, e.g. in version 2.x, since it's a breaking change.

**Reply by PsuFan** [reply]
I’d prefer for the Esp32 etc ports to assume ID 0 instead…

**Comment by robert-hh** [comment]
> the chip I am using: https://www.amazon.com/gp/product/B07X1W16QS/

According to the link, this is a standard ESP32, 4MB flash, NO SPIRAM. So the ESP GENERIC firmware should be the right one. Interestingly, the link says, the ESP32 is single/dual core. Ti tell the chip type, you have to look at the ESP32 chip on the board, It's printed on the package.

About I2C, I#ll ask my usual questions:
- Are pull-up resistors in place?
- Is the wiring complete and right? (GND, Vcc, SDA, SCL)
- If yo use a breadboard, could there be contact problems with the breadboard?
- You said you tested with Arduino. Did you use the same board for the test or an Arduino boars.
- Which mcp23017 set-up do you use? Is that a specific breakout board?

**Reply by PsuFan** [reply]
Hey robert, I didn't see where it mentioned no spiram, it says SRAM: 520 kB SRAM, not sure if that is the same thing.

I got the i2c working however with i2c = I2C(**0,** sda=Pin(21), scl=Pin(22))

**Reply by robert-hh** [reply]
It does not mention SPIRAM. It just says SRAM, which is the on-chip RAM. SPIRAM is an external SPI connected RAM chip.

**Comment by picpic020960** [comment]
Bonjour 
When generic spiram micropython build for ESP32-D0WDQ6 (V3) ?
I have I2C and SPI issues with twatch-2020 V2 (all is good with twatch-2020 V1)
Spiram is use with LVGL
Thanks for informations

**Reply by robert-hh** [reply]
In order to help you, more information is needed. Can you show the part of the code that sets up SPI/I2C and tries to use it. And which issues do you have. Do you get error messages? Is it simply not working?

**Reply by picpic020960** [reply]
Bonjour @robert-hh ,
precisions in : 
[https://github.com/russhughes/st7789_mpy/issues/89](url)
russhughes driver works but with 'flashing' and I don't know from where :
ESP32 version (V1/V3) , micropython firmware  , SPI /I2C from esp-idf ?
how to proceed for identify the root cause ?
Regards
the hardware is good : arduino/platformio firmware woks fine on V2

**Comment by picpic020960** [comment]
all in TTGO.h : translate in micropython initpower function and all is good : no noise on screen , i2c.scan() reliable
